### ISSUE TITLE
Object detection -- question about confidence measure

### ISSUE THREAD

**Issue by BERENZ** [apple/turicreate#2226]
I am running an object detection algorithm with custom classes and annotations, so I am using the object detection algorithm i.e. implementation of TinyYOLO.  `Predict` function returns column `confidence` but I have problem finding explanation what confidence actually means and how it is calculated (e.g. equations).  Is confidence just a probability that object is within given bounding box? 

Is it possible to calculate probability that a picture contains given object based solely on the object detection algorithm? If confidence is just a probability that object is within bounding box than it is easy, but if not then it may be more complicated. 

Something similar is done with image classification task but my problem requires approach with custom annotations as simple classification performs poorly. Thus, based on the object detection algorithm I would like to calculate probabilities whether object is in the picture.

**Comment by nickjong** [comment]
The confidence score is an output of the underlying neural network, so it is the result of training across a dataset where correct predictions are trained towards 1 and incorrect predictions are trained towards 0. If I'm not mistaken, the loss function used to train this output does treat it like a probability for the purpose of optimization, but your mileage may vary as to whether you should treat it as a literal probability.

**Comment by BERENZ** [comment]
Ok, so could you please point me the part where the prediction function is defined? 

Just to be sure, is the loss function defined the same as in the original YOLO [paper](https://pjreddie.com/media/files/papers/yolo_1.pdf) (page 4) or your reimplementation include some changes to the original algorithm?

**Comment by nickjong** [comment]
The code that transforms the raw forward predictions from the network to the human-readable bounding boxes is here: https://github.com/apple/turicreate/blob/b55af967aa2b824fdc37876e1fd7efebf44da0e1/src/python/turicreate/toolkits/object_detector/_detection.py#L92

The actual loss function is defined here: https://github.com/apple/turicreate/blob/b55af967aa2b824fdc37876e1fd7efebf44da0e1/src/python/turicreate/toolkits/object_detector/_mx_detector.py#L36

If I'm not mistaken, we're using YOLOv2, see also https://apple.github.io/turicreate/docs/userguide/object_detection/how-it-works.html

**Comment by BERENZ** [comment]
@nickjong  thank you for clarifying. I have one more question which is somehow connected to the initial question but I would like to be sure.  Does your implementation of YOLO 2 network is *the same* as the one provided in the [yolov2-tiny.cfg](https://github.com/AlexeyAB/darknet/blob/master/cfg/yolov2-tiny.cfg) and if not could you please indicate where are the differences?
